lib: fix `struct thread **` misuse in c-ares resolver bindings

eqvinox opened the pull request:
(commit message:)

The `struct thread **ref` that the thread code takes is written to and needs to stay valid over the lifetime of a thread.  This does not hold up if thread pointers are directly put in a `vector` since adding items to a `vector` may reallocate the entire array.  The thread code would then write to a now-invalid `ref`, potentially corrupting entirely unrelated data.

This should be extremely rare to trigger in practice since we only use one c-ares channel, which will likely only ever use one fd, so the vector is never resized.  That said, c-ares using only one fd is just plain fragile luck.

Either way, fix this by creating a resolver_fd tracking struct, and clean up the code while we're at it.

(this should be backported to 8.1 & 8.0 at least)

---

Randomly bumped into this while looking at the various uses the `vector` code has.  Turns out this one is plain wrong/buggy…

Added an "exercise" tool to run the resolver code while at it.  Refer to comment in `test_resolver.c` on why this isn't executed by `make check` or CI.

idryzhov commented on lib/resolver.c:
```diff
@@ -27,59 +28,119 @@ struct resolver_state {
+
+static struct resolver_fds_head resfds[1] = {INIT_HASH(resfds[0])};
+
+static struct resolver_fd *resolver_fd_get(int fd, struct resolver_state *state)
```
Could you rename the `state` argument? We already have a global variable named `state` so this can be misleading.

eqvinox replied:
Good point, hadn't noticed that.

idryzhov reviewed (approved):
Overall LGTM. I left just one comment in the code.
Also let's maybe remove `vector_get_index` API as it seems to be error-prone?
And also it looks like the change somehow affects `bgp_community_change_update`.

eqvinox commented:
> And also it looks like the change somehow affects bgp_community_change_update.

I have absolutely no idea (yet) what is going on there.  Looking into it…

eqvinox commented:
> Also let's maybe remove `vector_get_index` API as it seems to be error-prone?

ACK, after this (together with #9871) it's not used anymore and it's really a footgun.

idryzhov commented:
@eqvinox `bgp_community_change_update` was broken on master. Fixed already in https://github.com/FRRouting/frr/pull/10001 so you can rebase and it will be fine.

eqvinox commented:
> @eqvinox `bgp_community_change_update` was broken on master. Fixed already in #10001 so you can rebase and it will be fine.

Thanks for the pointer, rebased.  Also removed `vector_get_index()` now.
